sympy/core/add.py

```diff
@@ -1222,7 +1215,19 @@ def __neg__(self):
             return super().__neg__()
         return Add(*[-i for i in self.args])
 
-add = AssocOpDispatcher('add')
+
+def add(args, evaluate=False, **kwargs):
```

Comment:
There was a proposal some time ago (possibly some years ago) on the SymPy forum about modifying SymPy's behaviour so that `Add` would never evaluate, while the lowercase function (i.e. `add`) would perform the evaluation. The module `sympy.matrices.expressions` was written with that logic in mind (though no `add` function was added in the end, as no agreement was reached on how to implement it). In particular, you can notice that `MatAdd` never evaluates and requires you to call `MatAdd(X, X).doit()` in order to simplify the expression.

```python
In [2]: X = MatrixSymbol("X", 3, 3)

In [3]: MatAdd(X, X)
Out[3]: X + X

In [4]: MatAdd(X, X).doit()
Out[4]: 2*X
```

If we keep with this logic, why would you add an `evaluate=...` parameter to `add( ... )`?

Reply:
Removed evaluate option.

Reply:
The main part of that proposal was that `Add` should not evaluate which makes sense but `Add` still evaluates and this PR doesn't change that. As long as `Add` evaluates we need the option to call `Add` with `evaluate=False`. That means that if we are going to replace calls to `Add` with `add` then `add` needs to have at least the option to pass on `evaluate=False`.

Reply:
I think it is good to have `evaluate` parameter in `add` explicitly so that people can quickly see it with `help`. Also, since we are planning to encourage the users to use `add` instead of `Add`, letting default be `evaluate=False` may be better. @Upabjojr @oscarbenjamin how do you think?

Reply:
> since we are planning to encourage the users to use `add` instead of `Add`

Shouldn't they both be internal APIs? I suppose end-users should just use the plus sign `+`.

Reply:
End users need some way to create an unevaluated `Add`.

Reply:
`+`definitely should be the recommended way to get evaluated result, but it cannot pass any keyword argument. Sometimes the user might want to get unevaluated result, so I think both `+` and `add` should be exposed to end users.

I suggest that we follow `matrices` module's evaluation rule:
1. `a + b` always returns evaluated result.
2. `add([a, b])` returns unevaluated result.
3. `add([a, b], evaluate=True)` returns evaluated result.

Reply:
Point 2 may make rewriting rules a little harder. The idea of having both `add` and `Add` was to distinguish indeed the evaluated version (lowercase) and the unevaluated version (uppercase initial or camel case). Well, that was the idea in the original discussion we had on the forum some years ago, IIRC.

Suppose your expression is subject to some rewriting rules or substitution. For example, `(x+y).subs({x: y})`. I would expect to get `Mul(2, x)` instead of `Add(x, x)`. Currently, retriggering the expression construction and evaluation is performed by `expr.func(*expr.args)`. Would you need to add `evaluate=True` everywhere to keep the existing compatibility?

By digressing a bit, Wolfram Mathematica has a completely different strategy for unevaluated expressions. They have a `HoldForm` which stops the evaluation. However, in Python this would not be possible as the expression resolution always happens bottom up.

Reply:
In that idea is `add` a `Basic` subclass as an alternative to `Add`?

I agree that automatic evaluation is a mess right now but it's also somewhat orthogonal to the changes here. If you're concerned about the use of the name `add` then we could rename it to `_add`.

Reply:
@Upabjojr I am making `add` as a function (for user interface), not class. `(x+y).subs({x:y})` will call `Add(x, x)`, which will be evaluated to `Mul(2,x)`. I am not planning to change that behavior here.

Reply:
I have considered on this a bit, and concluded that `add` should have `evaluate=True` as default.

Reply:
In general I think that we should try to avoid introducing new automatic evaluation especially if `add` is going to be used internally by the core routines.

Can you explain the reasoning for having `add` evaluate by default?

Reply:
During the [discussion](https://github.com/sympy/sympy/pull/20659#discussion_r551287606) in previous PR, we have discovered the possibility to remove the subclasses of `Add`, `Mul` and `Pow` using kind system. That changed my idea about the purpose of `add`, `mul` and `power`. If we can remove the subclasses, then we can just use `Add` internally (which is preferable since we won't need to change every `Add` in the codebase to `add`). Then `add` will be a function for end users who will expect it to return evaluated result as default behavior.

Reply:
I don't expect many end users to use `add` rather than say `+`. We should assume that those who do use `Add` or `add` will do so because they know what they are doing and because it has some behaviour that they want.

We can't completely unify `MatAdd` and `Add` because they have different behaviour with respect to evaluation. I think in general in sympy that automatic evaluation is something of a mistake, especially when it takes place during `obj.func(*obj.args)` because it makes it impossible to work with "unevaluated" objects. Ideally we would change it so that `Add` does not evaluate by default but that's very difficult to change now.

If `add` is to be used by users and dispatch to both `MatAdd` and `Add` then we need to decide what it's behaviour is for evaluation now because it's difficult to change it later. On the other hand if `add` is introduced now as something internal/experimental and called e.g. `_add` then it would be easy to change its default later without worrying about backwards compatibility.

Reply:
If that's what it takes to remove the subclasses, I am interested in making `Add`, `Mul` and `Pow` unevaluated.

Reply:
Making Add, Mul and Pow unevaluated is an extremely big task and also a break in backwards compatibility. That's something that we would not be able to do without a big breaking release like sympy 2.0. I am potentially in favour of that but not for some time.

I suggest for now just keeping in mind the possibility of a change. I think it's more important to implement the kind system now rather than getting bogged down in a tangent.

Reply:
Okay. Then I'll introduce evaluated `add` for end users (so that we can have non-evaluated `Add` and evaluated `add` for sympy 2.0 in distant future), and unevaluated `_add` for internal use in this PR. @oscarbenjamin @Upabjojr , does this sound OK?

Reply:
> Making Add, Mul and Pow unevaluated is an extremely big task and also a break in backwards compatibility.

We could make them unevaluated, then provide a temporary wrapper calling `Add( ... ).doit()` and warning users that this behavior will be changed in the future.

Reply:
@Upabjojr  That's a good idea. Or we can raise a deprecation warning if `evaluate` option is not explicitly passed to `Add`. It still may be a big task, but I'm cut out for doing tedious work ;)

Reply:
@asmeurer I recall you mentioning this topic some while ago so I think that you might be interested in this discussion. Any comment?

Reply:
The place where we really need to fix this is not in what `Add(*args)` does because that can be controlled with `evaluate=False`. The real problem is there's no way to manipulate the "unevaluated" object if `obj.func(*obj.args)` triggers the evaluation. Changing the behaviour of `obj.func(*obj.args)` is needed so that you can use e.g. `subs` or even `xreplace` without triggering the evaluation:
```python
In [16]: p = Mul(Pow(sqrt(2), -1, evaluate=False), x, evaluate=False)

In [17]: p
Out[17]: 
x 
──
√2

In [18]: p.subs(x, y)
Out[18]: 
√2  
──⋅y
2   
```
I think making that change to the behaviour of `subs` or `xreplace` is desirable but it is very clearly not backwards compatible: users and downstream code expects the evaluation to occur when using `subs`.

There needs to be a bigger discussion about how evaluation should actually be controlled if it doesn't just happen automatically. I don't think that `doit` is a sufficient API. Proliferating calls to `doit` around your code will make everything slow. Also `doit` is not itself consistent within sympy.

At minimum we need to fix things like #17280 before making any kind of breaking change in evaluation.

Reply:
I see. I shall focus on getting this merged first, then. Will implementing non-evaluating `_add` make this PR approvable?

Reply:
> I don't think that `doit` is a sufficient API.

Would it help if `doit` were allowed to have an `include` and `exclude` option so, for example, if you wanted to do only one type of object you could do `expr.doit(Min)` or all but `Integral`, `expr.doit(exclude=Integral)`?